### PR TITLE
Fix various errors in examples

### PR DESCRIPTION
**Pull request by hannobraun**
I've been working on picking up the stalled #16. Since this PR includes a change request regarding an example, I wanted to make sure to properly test all of them, but found out that they're kind of a mess. There were a lot of build errors that a user copying them into a new project would face.

I fixed a lot of those errors, and left others in there. The commits have more details about that. I do think that this PR, while clearly an improvement, is just papering over the deeper issue: That the examples are ignored (as in, not compiled during `cargo test`), and therefore prone to break and become out of date. I think we should find a way to change that, but that is beyond the scope of this humble pull request.

### CODE REVIEW

**Comment by hannobraun** [comment]
I've pushed some additional fixes for problems that I found while working on #20. I've amended the existing commits, no new commits have been added.

File: src/lib.rs [hunk 1]
```diff
@@ -179,16 +190,21 @@
 //! implementation:
 //!
 //! ``` ignore
+//! extern crate embedded_hal as hal;
+//! extern crate nb;
+//!
 //! /// A synchronized serial interface
 //! // NOTE This is a global singleton
-//! pub struct Serial1;
+//! pub struct Serial1(Mutex<..>);
```
Comment (japaric) [inline src/lib.rs]: If this contains a field then it's no longer a global singleton. It's also not required to have this field. See below

File: src/lib.rs [hunk 2]
```diff
@@ -179,16 +190,21 @@
...
 //! static USART1: Mutex<_> = Mutex::new(..);
 //!
-//! impl hal::Serial for Serial {
-//!     fn read(&self) -> Result<u8, nb::Error<Error>> {
-//!         hal::Serial::read(&Serial(USART1.lock()))
+//! impl hal::serial::Read<u8> for Serial1 {
+//!     type Error = !;
+//!
+//!     fn read(&self) -> Result<u8, nb::Error<Self::Error>> {
+//!         hal::serial::Read::read(&Serial1(USART1.lock()))
```
Comment (japaric) [inline src/lib.rs]: This can use the `Serial` struct from before like this `hal::serial::Read::read(&Serial(&*USART1.lock()))`

File: src/lib.rs [hunk 3]
```diff
@@ -242,40 +258,46 @@
...
+//!     Future,
+//! };
+//! use futures::future::Loop;
 //! use stm32f103xx_hal_impl::{Serial, Timer};
 //!
 //! /// `futures` version of `Timer.wait`
 //! ///
 //! /// This returns a future that must be polled to completion
-//! fn wait<T>(timer: &T) -> impl Future<Item = (), Error = !>
+//! fn wait<T>(timer: T) -> impl Future<Item = (), Error = !>
```
Comment (japaric) [inline src/lib.rs]: With this signature the function will destroy the timer object. The pattern that the futures ecosystem uses to work around this issue is to return the input by value in `Item`. Thus the return type should be `impl Future<Item = T, Error = !>`.

File: src/lib.rs [hunk 4]
```diff
@@ -242,40 +258,46 @@
...
+//!     future::poll_fn(move || {
 //!         Ok(Async::Ready(try_nb!(timer.wait())))
 //!     })
 //! }
 //!
 //! /// `futures` version of `Serial.read`
 //! ///
 //! /// This returns a future that must be polled to completion
-//! fn read<S>(serial: &S) -> impl Future<Item = u8, Error = Error>
+//! fn read<S>(serial: S) -> impl Future<Item = u8, Error = S::Error>
```
Comment (japaric) [inline src/lib.rs]: same comment as above

File: src/lib.rs [hunk 5]
```diff
@@ -242,40 +258,46 @@
...
+//!     future::poll_fn(move || {
 //!         Ok(Async::Ready(try_nb!(serial.read())))
 //!     })
 //! }
 //!
 //! /// `futures` version of `Serial.write`
 //! ///
 //! /// This returns a future that must be polled to completion
-//! fn write<S>(byte: u8) -> impl Future<Item = (), Error = Error>
+//! fn write<S>(serial: S, byte: u8) -> impl Future<Item = (), Error = S::Error>
```
Comment (japaric) [inline src/lib.rs]: same comment as above

File: src/lib.rs (hunk 2 above)
Comment (hannobraun) [reply src/lib.rs]: Ah, got it. I missed the point of that example before.

File: src/lib.rs (hunk 3 above)
Comment (hannobraun) [reply src/lib.rs]: I don't think this pattern will work with `poll_fn`. Here's what I got now:
``` rust
fn wait<T>(timer: T) -> impl Future<Item = T, Error = !>
where
    T: hal::Timer,
{
    future::poll_fn(|| {
        try_nb!(timer.wait());
        Ok(Async::Ready(timer))
    })
}
```
And here's how it fails:
```
error[E0507]: cannot move out of captured outer variable in an `FnMut` closure
  --> src/lib.rs:29:25
   |
23 | fn wait<T>(timer: T) -> impl Future<Item = T, Error = !>
   |            ----- captured outer variable
...
29 |         Ok(Async::Ready(timer))
   |                         ^^^^^ cannot move out of captured outer variable in an `FnMut` closure
```
Which makes sense, as `poll_fn` takes an `FnMut` closure.
Of course I could re-create `poll_fn`/`PollFn`, but this seems a bit to involved for an example like this. What do you think?

**Comment by hannobraun** [comment]
Thanks for the review! I've pushed a commit addressing your first comment. I didn't adress your comments on the futures example (see the comment I left in reply to yours).

File: src/lib.rs (hunk 3 above)
Comment (japaric) [reply src/lib.rs]: Yeah, I *think* you'll have to use [loop_fn](https://docs.rs/futures/0.1.16/futures/future/fn.loop_fn.html) instead.

Comment (hannobraun) [reply src/lib.rs]: Thanks for the help! I think I got it now.

**Comment by hannobraun** [comment]
I've amended my last commit and force-pushed. It addresses all the review comments now.

In case the changes are okay: Do you want me to leave it like that, or should I squash the changes from the last commit into the original commits?

**Comment by japaric** [comment]
Thanks, @hannobraun.

@homunkulus r+

**Comment by homunkulus** [comment]
:pushpin: Commit dbafdc7 has been approved by `japaric`

**Comment by homunkulus** [comment]
:hourglass: Testing commit dbafdc7 with merge dbafdc7...

**Comment by homunkulus** [comment]
:sunny: Test successful - [status-travis](https://travis-ci.org/japaric/embedded-hal/builds/278732746?utm_source=github_status&utm_medium=notification)
Approved by: japaric
Pushing dbafdc79a0edfe38ea4797dcf62ef67c7f56b267 to master...

**Comment by hannobraun** [comment]
Thank you!

**Pull request merged**